File: scratch/generate_surahs.py
import re

# Existing file content
with open('../lib/core/models/surah.dart', 'r', encoding='utf-8') as f:
    old_content = f.read()

# Extract URLs
url_dict = {}
# simple extraction based on id
blocks = old_content.split('Surah(')
for block in blocks[1:]:
    # extract id
    id_match = re.search(r'id:\s*"([^"]+)"', block)
    url_match = re.search(r'url:\s*"([^"]+)"', block)
    if id_match and url_match:
        url_dict[id_match.group(1)] = url_match.group(1)

# Quran 114 metadata (Name, isMakki, English ID)
quran_meta = [
    ("الفاتحة", True, "al-fatiha"), ("البقرة", False, "al-baqara-22"), ("آل عمران", False, "al-imran"),
    ("النساء", False, "al-nisa"), ("المائدة", False, "al-maida"), ("الأنعام", True, "al-anam"),
    ("الأعراف", True, "al-araf"), ("الأنفال", False, "al-anfal"), ("التوبة", False, "al-tawba"),
    ("يونس", True, "yunus"), ("هود", True, "hud"), ("يوسف", True, "yusuf"),
    ("الرعد", False, "al-rad"), ("إبراهيم", True, "ibrahim"), ("الحجر", True, "al-hijr"),
    ("النحل", True, "al-nahl"), ("الإسراء", True, "al-isra"), ("الكهف", True, "al-kahf"),
    ("مريم", True, "maryam"), ("طه", True, "taha"), ("الأنبياء", True, "al-anbiya"),
    ("الحج", False, "al-hajj"), ("المؤمنون", True, "al-muminun"), ("النور", False, "al-nur"),
    ("الفرقان", True, "al-furqan"), ("الشعراء", True, "al-shuara"), ("النمل", True, "al-naml"),
    ("القصص", True, "al-qasas"), ("العنكبوت", True, "al-ankabut"), ("الروم", True, "al-rum"),
    ("لقمان", True, "luqman"), ("السجدة", True, "al-sajda"), ("الأحزاب", False, "al-ahzab"),
    ("سبأ", True, "saba"), ("فاطر", True, "fatir"), ("يس", True, "yaseen"),
    ("الصافات", True, "al-saffat"), ("ص", True, "sad"), ("الزمر", True, "al-zumar"),
    ("غافر", True, "ghafir"), ("فصلت", True, "fussilat"), ("الشورى", True, "al-shura"),
    ("الزخرف", True, "al-zukhruf"), ("الدخان", True, "al-dukhan"), ("الجاثية", True, "al-jathiya"),
    ("الأحقاف", True, "al-ahqaf"), ("محمد", False, "muhammad"), ("الفتح", False, "al-fath"),
    ("الحجرات", False, "al-hujurat"), ("ق", True, "qaf"), ("الذاريات", True, "al-zariyat"),
    ("الطور", True, "al-tur"), ("النجم", True, "al-najm"), ("القمر", True, "al-qamar"),
    ("الرحمن", False, "al-rahman"), ("الواقعة", True, "al-waqia-26"), ("الحديد", False, "al-hadid"),
    ("المجادلة", False, "al-mujadila"), ("الحشر", False, "al-hashr"), ("الممتحنة", False, "al-mumtahina"),
    ("الصف", False, "al-saff"), ("الجمعة", False, "al-juma"), ("المنافقون", False, "al-munafiqun"),
    ("التغابن", False, "al-taghabun"), ("الطلاق", False, "al-talaq"), ("التحريم", False, "al-tahrim"),
    ("الملك", True, "al-mulk"), ("القلم", True, "al-qalam"), ("الحاقة", True, "al-haqqa"),
    ("المعارج", True, "al-maarij"), ("نوح", True, "nuh"), ("الجن", True, "al-jin"),
    ("المزمل", True, "al-muzzammil"), ("المدثر", True, "al-muddaththir"), ("القيامة", True, "al-qiyama"),
    ("الإنسان", False, "al-insan"), ("المرسلات", True, "al-mursalat"), ("النبأ", True, "al-naba"),
    ("النازعات", True, "al-naziat"), ("عبس", True, "abasa"), ("التكوير", True, "al-takwir"),
    ("الإنفطار", True, "al-infitar"), ("المطففين", True, "al-mutaffifin"), ("الانشقاق", True, "al-inshiqaq"),
    ("البروج", True, "al-buruj"), ("الطارق", True, "al-tariq"), ("الأعلى", True, "al-ala"),
    ("الغاشية", True, "al-ghashiya"), ("الفجر", True, "al-fajr"), ("البلد", True, "al-balad"),
    ("الشمس", True, "al-shams"), ("الليل", True, "al-lail"), ("الضحى", True, "al-duha"),
    ("الشرح", True, "al-sharh"), ("التين", True, "al-tin"), ("العلق", True, "al-alaq"),
    ("القدر", True, "al-qadr"), ("البينة", False, "al-bayyina"), ("الزلزلة", False, "al-zalzala"),
    ("العاديات", True, "al-adiyat"), ("القارعة", True, "al-qaria"), ("التكاثر", True, "al-takathur"),
    ("العصر", True, "al-asr"), ("الهمزة", True, "al-humaza"), ("الفيل", True, "al-fil"),
    ("قريش", True, "quraish"), ("الماعون", True, "al-maun"), ("الكوثر", True, "al-kawthar"),
    ("الكافرون", True, "al-kafirun"), ("النصر", False, "al-nasr"), ("المسد", True, "al-masad"),
    ("الإخلاص", True, "al-ikhlas"), ("الفلق", True, "al-falaq"), ("الناس", True, "al-nas")
]

new_dart_code = """class Surah {
  final String id;
  final String name;
  final String url;
  final Duration estimatedDuration;
  final bool isMakki; // true for Makki, false for Madani

  Surah({
    required this.id,
    required this.name,
    required this.url,
    required this.estimatedDuration,
    required this.isMakki,
  });

  factory Surah.fromMap(Map<String, dynamic> map) {
    return Surah(
      id: map['id'],
      name: map['name'],
      url: map['url'],
      estimatedDuration: Duration(seconds: map['estimatedDuration']),
      isMakki: map['isMakki'] ?? true,
    );
  }
}

final List<Surah> surahList = [
"""

for name, is_makki, sid in quran_meta:
    url = url_dict.get(sid, "")
    makki_str = "true" if is_makki else "false"
    # Estimate duration: just a dummy 1200 if not found in old? No, we need something generic.
    # Al-Baqara is 4800, small are 60.
    dur = 1200
    if sid == "al-baqara-22": dur = 4800
    elif sid in ["al-imran", "al-nisa", "al-maida"]: dur = 3000
    new_dart_code += f"""  Surah(
    id: "{sid}",
    name: "سورة {name}",
    url: "{url}",
    estimatedDuration: Duration(seconds: {dur}),
    isMakki: {makki_str},
  ),
"""

new_dart_code += "];\n"

with open('../lib/core/models/surah.dart', 'w', encoding='utf-8') as f:
    f.write(new_dart_code)

File: scratch/generate_surahs.js
const fs = require('fs');

const oldContent = fs.readFileSync('../lib/core/models/surah.dart', 'utf-8');

const urlDict = {};
const blocks = oldContent.split('Surah(');
for (let i = 1; i < blocks.length; i++) {
    const idMatch = blocks[i].match(/id:\s*"([^"]+)"/);
    const urlMatch = blocks[i].match(/url:\s*"([^"]+)"/);
    if (idMatch && urlMatch) {
        urlDict[idMatch[1]] = urlMatch[1];
    }
}

const quranMeta = [
    ["الفاتحة", true, "al-fatiha"], ["البقرة", false, "al-baqara-22"], ["آل عمران", false, "al-imran"],
    ["النساء", false, "al-nisa"], ["المائدة", false, "al-maida"], ["الأنعام", true, "al-anam"],
    ["الأعراف", true, "al-araf"], ["الأنفال", false, "al-anfal"], ["التوبة", false, "al-tawba"],
    ["يونس", true, "yunus"], ["هود", true, "hud"], ["يوسف", true, "yusuf"],
    ["الرعد", false, "al-rad"], ["إبراهيم", true, "ibrahim"], ["الحجر", true, "al-hijr"],
    ["النحل", true, "al-nahl"], ["الإسراء", true, "al-isra"], ["الكهف", true, "al-kahf"],
    ["مريم", true, "maryam"], ["طه", true, "taha"], ["الأنبياء", true, "al-anbiya"],
    ["الحج", false, "al-hajj"], ["المؤمنون", true, "al-muminun"], ["النور", false, "al-nur"],
    ["الفرقان", true, "al-furqan"], ["الشعراء", true, "al-shuara"], ["النمل", true, "al-naml"],
    ["القصص", true, "al-qasas"], ["العنكبوت", true, "al-ankabut"], ["الروم", true, "al-rum"],
    ["لقمان", true, "luqman"], ["السجدة", true, "al-sajda"], ["الأحزاب", false, "al-ahzab"],
    ["سبأ", true, "saba"], ["فاطر", true, "fatir"], ["يس", true, "yaseen"],
    ["الصافات", true, "al-saffat"], ["ص", true, "sad"], ["الزمر", true, "al-zumar"],
    ["غافر", true, "ghafir"], ["فصلت", true, "fussilat"], ["الشورى", true, "al-shura"],
    ["الزخرف", true, "al-zukhruf"], ["الدخان", true, "al-dukhan"], ["الجاثية", true, "al-jathiya"],
    ["الأحقاف", true, "al-ahqaf"], ["محمد", false, "muhammad"], ["الفتح", false, "al-fath"],
    ["الحجرات", false, "al-hujurat"], ["ق", true, "qaf"], ["الذاريات", true, "al-zariyat"],
    ["الطور", true, "al-tur"], ["النجم", true, "al-najm"], ["القمر", true, "al-qamar"],
    ["الرحمن", false, "al-rahman"], ["الواقعة", true, "al-waqia-26"], ["الحديد", false, "al-hadid"],
    ["المجادلة", false, "al-mujadila"], ["الحشر", false, "al-hashr"], ["الممتحنة", false, "al-mumtahina"],
    ["الصف", false, "al-saff"], ["الجمعة", false, "al-juma"], ["المنافقون", false, "al-munafiqun"],
    ["التغابن", false, "al-taghabun"], ["الطلاق", false, "al-talaq"], ["التحريم", false, "al-tahrim"],
    ["الملك", true, "al-mulk"], ["القلم", true, "al-qalam"], ["الحاقة", true, "al-haqqa"],
    ["المعارج", true, "al-maarij"], ["نوح", true, "nuh"], ["الجن", true, "al-jin"],
    ["المزمل", true, "al-muzzammil"], ["المدثر", true, "al-muddaththir"], ["القيامة", true, "al-qiyama"],
    ["الإنسان", false, "al-insan"], ["المرسلات", true, "al-mursalat"], ["النبأ", true, "al-naba"],
    ["النازعات", true, "al-naziat"], ["عبس", true, "abasa"], ["التكوير", true, "al-takwir"],
    ["الإنفطار", true, "al-infitar"], ["المطففين", true, "al-mutaffifin"], ["الانشقاق", true, "al-inshiqaq"],
    ["البروج", true, "al-buruj"], ["الطارق", true, "al-tariq"], ["الأعلى", true, "al-ala"],
    ["الغاشية", true, "al-ghashiya"], ["الفجر", true, "al-fajr"], ["البلد", true, "al-balad"],
    ["الشمس", true, "al-shams"], ["الليل", true, "al-lail"], ["الضحى", true, "al-duha"],
    ["الشرح", true, "al-sharh"], ["التين", true, "al-tin"], ["العلق", true, "al-alaq"],
    ["القدر", true, "al-qadr"], ["البينة", false, "al-bayyina"], ["الزلزلة", false, "al-zalzala"],
    ["العاديات", true, "al-adiyat"], ["القارعة", true, "al-qaria"], ["التكاثر", true, "al-takathur"],
    ["العصر", true, "al-asr"], ["الهمزة", true, "al-humaza"], ["الفيل", true, "al-fil"],
    ["قريش", true, "quraish"], ["الماعون", true, "al-maun"], ["الكوثر", true, "al-kawthar"],
    ["الكافرون", true, "al-kafirun"], ["النصر", false, "al-nasr"], ["المسد", true, "al-masad"],
    ["الإخلاص", true, "al-ikhlas"], ["الفلق", true, "al-falaq"], ["الناس", true, "al-nas"]
];

let newCode = `class Surah {
  final String id;
  final String name;
  final String url;
  final Duration estimatedDuration;
  final bool isMakki;

  Surah({
    required this.id,
    required this.name,
    required this.url,
    required this.estimatedDuration,
    required this.isMakki,
  });

  factory Surah.fromMap(Map<String, dynamic> map) {
    return Surah(
      id: map['id'],
      name: map['name'],
      url: map['url'],
      estimatedDuration: Duration(seconds: map['estimatedDuration']),
      isMakki: map['isMakki'] ?? true,
    );
  }
}

final List<Surah> surahList = [
`;

for (const [name, isMakki, sid] of quranMeta) {
    const url = urlDict[sid] || "";
    const makkiStr = isMakki ? "true" : "false";
    let dur = 1200;
    if (sid === "al-baqara-22") dur = 4800;
    else if (["al-imran", "al-nisa", "al-maida"].includes(sid)) dur = 3000;
    
    newCode += `  Surah(
    id: "${sid}",
    name: "سورة ${name}",
    url: "${url}",
    estimatedDuration: Duration(seconds: ${dur}),
    isMakki: ${makkiStr},
  ),
`;
}

newCode += "];\n";

fs.writeFileSync('../lib/core/models/surah.dart', newCode, 'utf-8');
console.log("Done");
